noir/noir-repo/aztec_macros/src/transforms/event_interface.rs

```diff
@@ -0,0 +1,393 @@
+    // This did not work because I needed the submodule itself to add the trait impl back in to, but it would be nice if it was tagged on the module level
+    // let mut annotated_event_structs = module.submodules.iter_mut()
+    //     .flat_map(|submodule| submodule.contents.types.iter_mut())
+    //     .filter(|typ| typ.attributes.iter().any(|attr| is_custom_attribute(attr, "aztec(event)")));
+
+    // To diagnose
+    // let test = module.types.iter_mut();
+    // for event_struct in test {
+    //     print!("\ngenerate_event_interface_impl COUNT: {}\n", event_struct.name.0.contents);
+    // }
+
+    for submodule in module.submodules.iter_mut() {
+        let annotated_event_structs = submodule.contents.types.iter_mut().filter(|typ| {
+            typ.attributes.iter().any(|attr| is_custom_attribute(attr, "aztec(event)"))
+        });
+
+        for event_struct in annotated_event_structs {
+            // event_struct.attributes.push(SecondaryAttribute::Abi("events".to_string()));
+            // If one impl is pushed, this doesn't throw the "#[abi(tag)] attributes can only be used in contracts" error
+            // But if more than one impl is pushed, we get an increasing amount of "#[abi(tag)] attributes can only be used in contracts" errors
+            // We work around this by doing this addition in the HIR pass via transform_event_abi below.
+
+            let event_type = event_struct.name.0.contents.to_string();
+            let event_len = event_struct.fields.len() as u32;
+            // event_byte_len = event fields * 32 + randomness (32) + event_type_id (32)
+            let event_byte_len = event_len * 32 + 64;
+
+            let mut event_fields = vec![];
+
+            for (field_ident, field_type) in event_struct.fields.iter() {
+                event_fields.push((
+                    field_ident.0.contents.to_string(),
+                    field_type.typ.to_string().replace("plain::", ""),
+                ));
+            }
+
+            let mut event_impl = generate_impl_stub_event(event_type.as_str())?;
+            event_impl
+                .methods
+                .push((generate_fn_selector(event_type.as_str(), event_len)?, event_struct.span));
+            submodule.contents.impls.push(event_impl);
+
+            let mut event_interface_trait_impl =
+                generate_trait_impl_stub_event_interface(event_type.as_str(), event_byte_len)?;
+            event_interface_trait_impl.items.push(TraitImplItem::Function(
+                generate_fn_get_event_type_id(event_type.as_str(), event_len)?,
+            ));
+            event_interface_trait_impl.items.push(TraitImplItem::Function(
+                generate_fn_private_to_be_bytes(event_type.as_str(), event_byte_len)?,
+            ));
+            // This does not work because of a weird noir bug, where two generics are not allowed
```

Comment:
Ah, this was the issue from earlier. Is there a noir issue for this? Do they know about the issue?